### ISSUE TITLE
JDK18 aarch64_mac jniargtests_2_FAILED - JNI ARG Tests FAILED

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse-openj9/openj9#14500]
Failure link
------------

From an internal build `job/Test_openjdk18_j9_sanity.functional_aarch64_mac_testList_0/3/`(`macaarch64rt2`):
```
openjdk version "18" 2022-03-22
IBM Semeru Runtime Open Edition 18.0.0.0-m1 (build 18+35)
Eclipse OpenJ9 VM 18.0.0.0-m1 (build openj9-0.31.0-m1, JRE 18 Mac OS X aarch64-64-Bit 20220210_8 (JIT enabled, AOT enabled)
OpenJ9   - 700cde7b8
OMR      - bb5d44b8c
JCL      - cdd4205c649 based on jdk-18+35)
```
[Rerun in Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?SDK_RESOURCE=upstream&amp;TARGET=jniargtests_2&amp;TEST_FLAG=&amp;UPSTREAM_TEST_JOB_NAME=Test_openjdk18_j9_sanity.functional_aarch64_mac&amp;DOCKER_REQUIRED=false&amp;ACTIVE_NODE_TIMEOUT=0&amp;VENDOR_TEST_DIRS=&amp;EXTRA_DOCKER_ARGS=&amp;TKG_OWNER_BRANCH=adoptium%3Amaster&amp;OPENJ9_SYSTEMTEST_OWNER_BRANCH=eclipse%3Amaster&amp;PLATFORM=aarch64_mac&amp;GENERATE_JOBS=false&amp;KEEP_REPORTDIR=false&amp;PERSONAL_BUILD=false&amp;ADOPTOPENJDK_REPO=https%3A%2F%2Fgithub.com%2Fadoptium%2Faqa-tests.git&amp;LABEL=&amp;EXTRA_OPTIONS=&amp;CUSTOMIZED_SDK_URL=+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-mac-aarch64-openj9%2F8%2Fibm-semeru-open-jdk_aarch64_mac_18_35_openj9-0.31.0-m1.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-mac-aarch64-openj9%2F8%2Fibm-semeru-open-debugimage_aarch64_mac_18_35_openj9-0.31.0-m1.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-mac-aarch64-openj9%2F8%2Fibm-semeru-open-testimage_aarch64_mac_18_35_openj9-0.31.0-m1.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk18%2Fjdk18-mac-aarch64-openj9%2F8%2Fibm-semeru-open-jre_aarch64_mac_18_35_openj9-0.31.0-m1.tar.gz&amp;BUILD_IDENTIFIER=&amp;ADOPTOPENJDK_BRANCH=master&amp;LIGHT_WEIGHT_CHECKOUT=true&amp;USE_JRE=false&amp;ARTIFACTORY_SERVER=na.artifactory.swg-devops&amp;KEEP_WORKSPACE=false&amp;USER_CREDENTIALS_ID=&amp;JDK_VERSION=18&amp;ITERATIONS=1&amp;VENDOR_TEST_REPOS=&amp;JDK_REPO=https%3A%2F%2Fgithub.com%2Fibmruntimes%2Fopenj9-openjdk-jdk18&amp;RELEASE_TAG=v0.31.0-release&amp;OPENJ9_BRANCH=v0.31.0-release&amp;OPENJ9_SHA=&amp;JCK_GIT_REPO=&amp;VENDOR_TEST_BRANCHES=&amp;OPENJ9_REPO=https%3A%2F%2Fgithub.com%2Feclipse-openj9%2Fopenj9.git&amp;UPSTREAM_JOB_NAME=&amp;CLOUD_PROVIDER=&amp;CUSTOM_TARGET=&amp;VENDOR_TEST_SHAS=&amp;JDK_BRANCH=v0.31.0-release&amp;LABEL_ADDITION=&amp;ARTIFACTORY_REPO=&amp;ARTIFACTORY_ROOT_DIR=&amp;UPSTREAM_TEST_JOB_NUMBER=3&amp;DOCKERIMAGE_TAG=&amp;JDK_IMPL=openj9&amp;TEST_TIME=&amp;SSH_AGENT_CREDENTIAL=83181e25-eea4-4f55-8b3e-e79615733226&amp;AUTO_DETECT=true&amp;SLACK_CHANNEL=&amp;DYNAMIC_COMPILE=false&amp;ADOPTOPENJDK_SYSTEMTEST_OWNER_BRANCH=adoptium%3Amaster&amp;CUSTOMIZED_SDK_URL_CREDENTIAL_ID=4e18ffe7-b1b1-4272-9979-99769b68bcc2&amp;ARCHIVE_TEST_RESULTS=false&amp;NUM_MACHINES=&amp;OPENJDK_SHA=&amp;TRSS_URL=http%3A%2F%2Ftrss1.fyre.ibm.com&amp;BUILD_LIST=functional&amp;USE_TESTENV_PROPERTIES=false&amp;UPSTREAM_JOB_NUMBER=&amp;STF_OWNER_BRANCH=adoptium%3Amaster&amp;TIME_LIMIT=20&amp;JVM_OPTIONS=&amp;PARALLEL=None) - Change TARGET to run only the failed test targets.

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------
```
[2022-02-11T05:25:51.716Z] variation: -Xjit:count=1,disableAsyncCompilation
[2022-02-11T05:25:51.716Z] JVM_OPTIONS:  -Xjit:count=1,disableAsyncCompilation 

[2022-02-11T05:26:02.503Z] Argument error:  nativeIIIIIZZZZZrZ::arg8 got: 00000000 expected: 00000001
[2022-02-11T05:26:02.503Z] Argument error:  nativeIIIIIZZZZZrZ::arg10 got: 00000000 expected: 00000001
[2022-02-11T05:26:02.503Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg17 got: 00000000 expected: 00000002
[2022-02-11T05:26:02.503Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg18 got: 00000000 expected: 00000103
[2022-02-11T05:26:02.503Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg19 got: 00000002 expected: 00000104
[2022-02-11T05:26:02.503Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg20 got: 00000103 expected: BEEF0005
[2022-02-11T05:26:02.503Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg21 got: BEEF000500000104 expected: FEED0606FACE0006
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg22 got: FACE0006 expected: 418D999A
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg23 got: 00000000418D999A expected: 4032CCCCCCCCCCCD
[2022-02-11T05:26:02.504Z] Argument error:  nativeIIIIIZZZZZrZ::arg8 got: 00000000 expected: 00000001
[2022-02-11T05:26:02.504Z] Argument error:  nativeIIIIIZZZZZrZ::arg10 got: 00000000 expected: 00000001
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg17 got: 00000000 expected: 00000002
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg18 got: 00000000 expected: 00000103
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg19 got: 00000002 expected: 00000104
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg20 got: 00000103 expected: BEEF0005
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg21 got: BEEF000500000104 expected: FEED0606FACE0006
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg22 got: FACE0006 expected: 418D999A
[2022-02-11T05:26:02.504Z] Argument error:  nativeBBSSIJFDIFDFDFDBBSSIJFDrJ::arg23 got: 40361EB8418D999A expected: 4032CCCCCCCCCCCD
[2022-02-11T05:26:02.504Z] Argument error:  nativeIIIIIZZZZZrZ::arg8 got: 00000000 expected: 00000001
[2022-02-11T05:26:02.504Z] Argument error:  nativeIIIIIZZZZZrZ::arg10 got: 00000000 expected: 00000001
[2022-02-11T05:26:02.504Z]   JNI Tests: 1356
[2022-02-11T05:26:02.504Z]   Argument Failures: 20
[2022-02-11T05:26:02.504Z]   Return Value Failures: 0
[2022-02-11T05:26:02.504Z] JNI ARG Tests FAILED
[2022-02-11T05:26:02.504Z] 
[2022-02-11T05:26:02.504Z] jniargtests_2_FAILED
```

**Comment by pshipton** [comment]
@knn-k

**Comment by knn-k** [comment]
PR #14479  will resolve the failures with `nativeBBSSIJFDIFDFDFDBBSSIJFDrJ`.  I am investigating the failures with `nativeIIIIIZZZZZrZ` now.

**Comment by knn-k** [comment]
It was the change in #4569 that caused the failures with `nativeIIIIIZZZZZrZ`.

The type of `arg6` - `arg10` is `jboolean` in the JNI function declaration. 
https://github.com/eclipse-openj9/openj9/blob/ce1c984f0653e047c22eb8718a6f53e040500300/runtime/tests/jniarg/jniargtests.h#L518

The JNI code expects those arguments to be `signed char`, while and the VM interpreter passes them as `uint32` to the JNI function.  This inconsistency caused the failures on AArch64 macOS.

It doesn't matter on AArch64 Linux, for example, because all the arguments passed in the C stack are aligned at 8-byte boundary, and the platform endian is little.  On the other hand, the alignment of the arguments in the C stack is different from type to type on AArch64 macOS.

**Comment by knn-k** [comment]
See also the discussion in #14560 for the `boolean` -> `uint32` conversion by the VM interpreter.  I am going to disable it.

**Comment by knn-k** [comment]
I opened Issue #14574 for further discussion on the boolean type conversion.

**Comment by knn-k** [comment]
All the jniargtests pass on AArch64 macOS when #14575 is merged.

**Comment by knn-k** [comment]
The fix was also merged to the v0.31.0 branch as #14603.  I think we can close this issue after the next run of sanity.functional.

**Comment by pshipton** [comment]
Rather than try to remember to close it, I'll just close it now and we can reopen if the problem is seen again.
